Mathlib/Tactic/Linter/TextBased.lean

```diff
@@ -338,9 +350,39 @@ def broadImportsLinter : TextbasedLinter := fun lines ↦ Id.run do
             else if name == "Lake" || name.startsWith "Lake." then
               errors := errors.push (StyleError.broadImport BroadImports.Lake, lineNumber)
       lineNumber := lineNumber + 1
-  return errors
+  return (errors, none)
 
 
+/-- Lint a collection of input strings if one of them contains windows line endings. -/
+def windowsLineEndingLinter : TextbasedLinter := fun lines ↦ Id.run do
+  let mut errors := Array.mkEmpty 0
+  let mut fixedLines := lines
+  -- invariant: this equals the current index of 'line' in 'lines',
+  -- hence starts at 0 and is incremented *at the end* of the loop
+  let mut lineNumber := 0
+  for line in lines do
+    let replaced := line.crlfToLf
+    if replaced != line then
+      errors := errors.push (StyleError.windowsLineEnding, lineNumber)
+      fixedLines := fixedLines.set! 0 replaced
+    lineNumber := lineNumber + 1
+  return (errors, fixedLines)
+
+
+/-- Lint a collection of input strings if one of them contains trailing whitespace. -/
+def trailingWhitespaceLinter : TextbasedLinter := fun lines ↦ Id.run do
+  let mut errors := Array.mkEmpty 0
+  let mut fixedLines := lines
+  -- invariant: this equals the current index of 'line' in 'lines',
+  -- hence starts at 0 and is incremented *at the end* of the loop
+  let mut lineNumber := 0
+  for line in lines do
+    if line.back == ' ' then
```

Comment:
```suggestion
    if line.back == ' ' then
```
You may be able to use `String.trimRight` and completely copy over the previous one.  In fact, if there are more than two of these "line-by-line" replacements, it may make sense to abstract "the looping over the lines and bookkeep the change" logic.

Reply:
Yes, I could do so, the result should be equivalent. In general, I'd like to avoid making such a "no-op" write for every line in the file, for every linter with an auto-fix (as I imagine that to be slower than necessary. `trimRight` or `dropWhileRight` seem to do more work than simply checking the last character: in the common case, no action should be needed.

That said: let me try to actually measure this.


. I'm a bit cautious to do this in general --- but perhaps that's premature optimisation.

Reply:
I did a quick benchmark (disabling the Python linters, and running the text-based ones a few times in a row). Always overwriting consistently produced a measurable slow-down, of 2,4 -> 2,55s. It appears the trailing whitespace linter takes about 50ms as-is, and perhaps 200ms afterwards. That doesn't strike me as a good change.

Reply:
(Side note: the windows line endings (550ms) and adaptation note linters (800ms) are by far the slowest linters currently. Lean's string matching is not as optimised, it seems.)

Reply:
Let me resolve this comment - if you disagree, let's discuss!